### ISSUE TITLE
Merge sub-crates to the crossterm crate

### ISSUE THREAD

**Issue by zrzka** [crossterm-rs/crossterm#265]
## Goal

Provide all the features we do provide now through the `crossterm` crate only. Archive sub-crates and inform all users.

This does not include `crossterm_winapi`, which can still be used by anyone outside the `crossterm` ecosystem.

We will end up with three active repositories:

* `crossterm-rs/crossterm`
* `crossterm-rs/crossterm-winapi`
* `crossterm-rs/examples`

## Motivation

There're dependencies between sub-crates we would like to avoid. Some of them are also using `sys` module, which needs to be public, but it shouldn't be. It's not part of the public API contract.

There're inconsistencies in examples, book, readmes, ... Some of the use `crossterm` crate with features, some of the use `crossterm-*` sub-crates, ... User can be puzzled what to use. In fact, it doesn't matter what the user will use, but the question is why we do provide two ways to achieve same thing.

And the last one is `crossterm_utils`, which is published as a crate, but it's not useful outside of them `crossterm` ecosystem.

## Timeframe

There's no date set. I would like to focus on phase 1 now to finish the cleanup process I started. Once I finish it, I'll focus on bugs. Then, can be even distant future, discuss with Timon if it's the right time to merge.

## Phase 1 - Cleanup

Improve & polish crates public (docs.rs & README) documentation (@zrzka):

- [x] `cursor` - https://github.com/crossterm-rs/crossterm-cursor/pull/2 & https://github.com/crossterm-rs/crossterm-cursor/pull/7
- [x] `terminal` - https://github.com/crossterm-rs/crossterm-terminal/pull/3
- [x] `screen` - https://github.com/crossterm-rs/crossterm-screen/pull/5
- [x] `style` - https://github.com/crossterm-rs/crossterm-style/pull/2 & https://github.com/crossterm-rs/crossterm-style/pull/3
- [x] `input` - https://github.com/crossterm-rs/crossterm-input/pull/3 & https://github.com/crossterm-rs/crossterm-input/pull/4
- [x] `utils` - https://github.com/crossterm-rs/crossterm-utils/pull/3
- [x] `crossterm` - https://github.com/crossterm-rs/crossterm/pull/269
- [x] Publish new book with info that the book is deprecated & no longer maintained

This includes copying [the book](https://github.com/crossterm-rs/crossterm/tree/master/docs/mdbook) chapters to their respective `lib.rs` modules.

## Phase 2 - Book removal

Then remove the book references from:

- [x] `cursor` - https://github.com/crossterm-rs/crossterm-cursor/pull/8
- [x] `terminal` - https://github.com/crossterm-rs/crossterm-terminal/pull/4
- [x] `screen` - https://github.com/crossterm-rs/crossterm-screen/pull/6
- [x] `style` - https://github.com/crossterm-rs/crossterm-style/pull/4
- [x] `input`- https://github.com/crossterm-rs/crossterm-input/pull/6
- [x] `utils`- https://github.com/crossterm-rs/crossterm-utils/pull/4
- [x] `crossterm` (& the book itself) - https://github.com/crossterm-rs/crossterm/pull/270

## Phase 3 - Bug fixing

Fix known bugs. The motivation here is to avoid archiving (GH) / deprecating (crates.io) buggy crates. Not all bugs can be fixed, especially those we don't know about, but rest should be fixed.

#### Windows related

- [x] Make read sync block for windows systems
     - Issue https://github.com/crossterm-rs/crossterm/issues/195
     - PR https://github.com/crossterm-rs/crossterm-winapi/pull/2
- [x] Sync raw mode behavior
     - Issue https://github.com/crossterm-rs/crossterm/issues/250
     - PR https://github.com/crossterm-rs/crossterm-screen/pull/3
- [x] Sync grey/white color
    - Issue https://github.com/crossterm-rs/crossterm/issues/263
    - PR https://github.com/crossterm-rs/crossterm-style/pull/5

#### All platforms

- [x] Add Commands: EnterAlternateScreen/LeaveAlternateScreen
     - PR https://github.com/crossterm-rs/crossterm-screen/pull/7
- [x] Sync mouse button coordinates
    - Issue https://github.com/crossterm-rs/crossterm/issues/267
    - PR https://github.com/crossterm-rs/crossterm-input/pull/7
- [x] Remove unsafe
    - Issue https://github.com/crossterm-rs/crossterm/issues/245
    - PRs https://github.com/crossterm-rs/crossterm-utils/pull/5 & https://github.com/crossterm-rs/crossterm-cursor/pull/9
- [x] Reexport `ResetColor`
    - Issue https://github.com/crossterm-rs/crossterm/issues/273
    - PR https://github.com/crossterm-rs/crossterm/pull/274

#### UNIX

- [x] Async reader doesn't produce MouseEvents and cursor position is swallowed by the async reader
    - Issue https://github.com/crossterm-rs/crossterm/issues/271
    - Issue https://github.com/crossterm-rs/crossterm/issues/199
    - Issue https://github.com/crossterm-rs/crossterm/issues/253
    - PR https://github.com/crossterm-rs/crossterm-input/pull/9

### Reminder

Just a check to go through the issues again if there isn't any new critical issue.

- [x] Known bugs fixed and we're happy to proceed with phase 4

### For every sub-crate:

* Add a note to the README that the crate is deprecated and no longer maintained.
* Add a note to the `lib.rs` that the crate is deprecated and no longer maintained.
* Bump patch in the `Cargo.toml` and add `[badges] maintenance = { status = "deprecated" }`
* Update CHANGELOG.md with the same info
* Release
* Archive on the GitHub

- [x] crossterm_winapi
  - PR https://github.com/crossterm-rs/crossterm-winapi/pull/4
- [x] crossterm_utils
  - PR https://github.com/crossterm-rs/crossterm-utils/pull/7
- [x] crossterm_screen
  - PR https://github.com/crossterm-rs/crossterm-screen/pull/9
- [x] crossterm_style
  - PR https://github.com/crossterm-rs/crossterm-style/pull/6
- [x] crossterm_input
  - PR https://github.com/crossterm-rs/crossterm-input/pull/10
- [x] crossterm_cursor
  - PR https://github.com/crossterm-rs/crossterm-cursor/pull/11
- [x] crossterm_terminal
  - PR https://github.com/crossterm-rs/crossterm-terminal/pull/5

### `crossterm` crate

- [x] Release new `crossterm` crate with updated dependencies
  - PR https://github.com/crossterm-rs/crossterm/pull/277

By the end of this phase, all sub-crates are polished, hopefully, bug-free and marked as deprecated.

### For every sub-crate

Stop accepting sub-crates PRs (already archived) and copy the code to the `crossterm` crate. This step requires (for every sub-crate):

* copy the code to the respective `crossterm` module (`crossterm_cursor` -> `cursor`)
* feature gate the module and root re-exports
* fix docstring docs/examples (`use crossterm_cursor` -> `crossterm`)
* fix paths (`crate` will differ, etc.)
* update features (if there's a sub-crate dependency, like `input` requires `screen`, the `input` feature must enable `screen` feature as well)

- [x] `utils`
- [x] `screen`
- [x] `style`
- [x] `input`
- [x] `cursor`
- [x] `terminal`

### `crossterm` crate

Once everything is merged, documentation is updated, etc.

- [x] Add a note to the README that all the subcrates are deprecated/no longer maintained
- [x] Merge the old command API docs & emphasis that it's recommended to use it
- [x] Bump patch and release new `crossterm` crate version

## Phase 6 - Inform users

- [x] Discord
- [x] Broot - https://github.com/Canop/broot/issues/70

### Not need to inform the following projects

All of them are using `crossterm` crate directly.

- Cursive - https://github.com/gyscos/Cursive
- TUI - https://github.com/fdehau/tui-rs
- Rust-sloth - https://github.com/jonathandturner/rust-sloth/tree/crossterm-port

**Comment by TimonPost** [comment]
- [ ] read sync: 
     - PR https://github.com/crossterm-rs/crossterm-winapi/pull/2
     - Issue: https://github.com/crossterm-rs/crossterm/issues/195
- [ ] raw mode 
     - https://github.com/crossterm-rs/crossterm/issues/250
     - https://github.com/crossterm-rs/crossterm-screen/pull/3
- [ ] Add Commands: EnterAlternateScreen/LeaveAlternateScreen
- [ ] sync mouse button coordinates: https://github.com/crossterm-rs/crossterm/issues/267
- [ ] sync grey/white color: https://github.com/crossterm-rs/crossterm/issues/263

**Comment by Canop** [comment]
Beautiful!

**Comment by TimonPost** [comment]
Everyone that contributed to this version thanks! https://crates.io/crates/crossterm/0.12.1
